### PR TITLE
Split off watcher from develop, add orchestration to main

### PR DESCRIPTION
**Pull request by cecyc**
Part of https://github.com/Shopify/shopify-cli-extensions/issues/2

* splits `build.Develop` into `build.Develop` and `build.Watch`
* refactor `build.Build`, `build.Develop`, and `build.Watch` to not return anything, but yield to callback function
* add goroutine and channel orchestration for `build.Develop` and `build.Watch` to main
* adds `Extension` to `build.Builder` interface to be used to differentiate processes by UUID
* removes `build/testdata/checkout-ui-extension` as it is unused

**Notice the `NOTIFY` log line, which shows we're hitting `api.Notify`

https://user-images.githubusercontent.com/989784/131912369-6d3a9ffe-1f39-4a60-9334-4124a8557e37.mp4

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -62,6 +62,7 @@ func (api *ExtensionsApi) PrintOutStatus() {
 }
 
 func (api *ExtensionsApi) Notify(statusUpdate StatusUpdate) error {
+	log.Print("***NOTIFY EVENT")
```
Comment (cecyc) [inline api/api.go:65]: I'll remove this log line once we have successfully tophatted

File: main.go [hunk 2]
```diff
@@ -43,19 +43,31 @@ type CLI struct {
...
-			log.Printf("Extension %s built successfully!", e.UUID)
-		} else {
-			build_errors += 1
-			log.Printf("Extension %s failed to build: %s", e.UUID, result.Error.Error())
+		go b.Build(ctx, func(result build.Result) {
+			ch <- result
+		})
+
+		for result := range ch {
+			active_builders--
```
Comment (t6d) [inline main.go:59]: Did you see my comment regarding [WaitGroups](https://gobyexample.com/waitgroups) earlier and how that could resolve manually keeping track of counters?

**Review by t6d — APPROVED** [review]
Using wait groups can be shipped in a subsequent PR and is not high priority. Code looks good to me, thanks for the tophat video!

File: main.go [hunk 3]
```diff
@@ -80,5 +92,49 @@ func (cli *CLI) create(args ...string) {
...
+	<-watch_chan
+}
+
+func (cli *CLI) monitor(active_workers int, ch chan build.Result, action string, a *api.ExtensionsApi) {
+	for result := range ch {
+		active_workers--
+
+		if result.Success {
+			log.Printf("[%s] event for extension: %s", action, result.UUID)
+			go a.Notify(api.StatusUpdate{Type: "Success", Extensions: cli.config.Extensions})
```
Comment (vividviolet) [inline main.go:130]: Can we notify only with the updated extension? This way the Dev Console can update the status for the right extension

File: main.go [hunk 4]
```diff
@@ -80,5 +92,49 @@ func (cli *CLI) create(args ...string) {
...
+		} else {
+			log.Printf("[%s] error for extension %s, error: %s", action, result.UUID, result.Error.Error())
+			go a.Notify(api.StatusUpdate{Type: "Error", Extensions: cli.config.Extensions})
```
Comment (vividviolet) [inline main.go:133]: Same here, it should report the status for only the extension that has the error

**Review by vividviolet — COMMENTED** [review]
Nice job! One small request and the it's good to go

File: main.go (hunk 3 above)
Comment (cecyc) [reply main.go]: Done! ✅

File: main.go (hunk 4 above)
Comment (cecyc) [reply main.go]: Done! ✅

**Pull request merged**